File: examples/relax_c/relax.c
/*
 * One VASP relaxation, authored in C: the same three-step shape as the Bash
 * runner httk.workflow.vasp.runners/vasp_relax.sh, built on the native C SDK.
 *
 *   prepare  stage the payload POSCAR (and INCAR if present) into the workdir
 *   run      run the configured VASP command and classify what it did
 *   publish  copy the finished calculation into the job's transactional data
 *
 * Every httk_workflow_* call reaches the same implementation the Python and Bash
 * SDKs do, so this runner is mock-vasp compatible and publishes the same bytes.
 * Build it with the Makefile beside this file:
 *
 *     make
 *
 * See ../mock_vasp.py for a stand-in VASP, and README.md for the whole flow.
 */

#define _POSIX_C_SOURCE 200809L

#include "httk_workflow.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/stat.h>

/* The files a finished relaxation publishes, if the run produced them. */
static const char *const COLLECT[] = {
    "INCAR", "KPOINTS", "OUTCAR", "CONTCAR", "OSZICAR", "vasprun.xml", "vasp-run-report.json", NULL,
};

static int file_exists(const char *path) {
    struct stat info;
    return stat(path, &info) == 0 && S_ISREG(info.st_mode);
}

/* Join two path components; the caller frees. */
static char *join_path(const char *a, const char *b) {
    size_t need = strlen(a) + 1 + strlen(b) + 1;
    char *out = malloc(need);
    if (out != NULL) {
        snprintf(out, need, "%s/%s", a, b);
    }
    return out;
}

/* Copy one file byte for byte; returns 0 on success. */
static int copy_file(const char *source, const char *destination) {
    FILE *in = fopen(source, "rb");
    if (in == NULL) {
        return -1;
    }
    FILE *out = fopen(destination, "wb");
    if (out == NULL) {
        fclose(in);
        return -1;
    }
    char buffer[8192];
    size_t got;
    int status = 0;
    while ((got = fread(buffer, 1, sizeof buffer, in)) > 0) {
        if (fwrite(buffer, 1, got, out) != got) {
            status = -1;
            break;
        }
    }
    if (fclose(out) != 0) {
        status = -1;
    }
    fclose(in);
    return status;
}

/* Stage a payload-relative file named by one parameter into the workdir. */
static int stage_input(const char *job_dir, const char *parameter, const char *fallback, const char *destination) {
    int status;
    char *relative = httk_workflow_parameter(parameter, fallback, &status);
    if (relative == NULL) {
        return -1;
    }
    char *source = join_path(job_dir, relative);
    free(relative);
    if (source == NULL) {
        return -1;
    }
    int result = -1;
    if (file_exists(source)) {
        result = copy_file(source, destination) == 0 ? 1 : -1;
    } else {
        result = 0; /* not present */
    }
    free(source);
    return result;
}

static int step_prepare(void) {
    const char *job_dir = getenv("HTTK_WORKFLOW_JOB_DIR");
    if (job_dir == NULL) {
        job_dir = ".";
    }

    int staged = stage_input(job_dir, "poscar", "files/POSCAR", "POSCAR");
    if (staged <= 0) {
        httk_workflow_fail("vasp.input_missing", "the starting structure is not in this payload", NULL);
        return 0;
    }
    /* An INCAR is optional; the mock VASP reads only the POSCAR. */
    stage_input(job_dir, "incar", "files/INCAR", "INCAR");

    httk_workflow_runlog_note("prepared a relaxation");
    httk_workflow_advance("run", NULL);
    return 0;
}

static int step_run(void) {
    int status;
    char *from_parameter = httk_workflow_parameter("vasp_command", "", &status);
    char *command = httk_workflow_setting("vasp.command", from_parameter != NULL ? from_parameter : "", &status);
    free(from_parameter);
    if (command == NULL || *command == '\0') {
        free(command);
        httk_workflow_fail("vasp.command_missing",
                           "no VASP command is configured: set it with "
                           "httk workflow workspace settings set vasp.command '...', or set "
                           "HTTK_VASP_COMMAND, or give the job a vasp_command parameter",
                           NULL);
        return 0;
    }

    char *timeout = httk_workflow_parameter("timeout", "86400", &status);

    /* The resolved command is one argv string; split it on whitespace, the way
     * the Bash runner leaves it unquoted for the shell to word-split. There can
     * be no more tokens than characters, so size the array for the worst case
     * once and never grow (nor silently truncate) it. */
    const char *fixed[] = {"--timeout", timeout != NULL ? timeout : "86400", "--report", "vasp-run-report.json", "--"};
    size_t fixed_n = sizeof fixed / sizeof fixed[0];
    size_t capacity = fixed_n + strlen(command) + 1;
    const char **args = malloc((capacity + 1) * sizeof *args);
    size_t n = 0;
    if (args != NULL) {
        for (size_t i = 0; i < fixed_n; i++) {
            args[n++] = fixed[i];
        }
        for (char *token = strtok(command, " \t"); token != NULL; token = strtok(NULL, " \t")) {
            args[n++] = token;
        }
        args[n] = NULL;
    }

    int run_status = args != NULL ? httk_workflow_run(args) : HTTK_WORKFLOW_REFUSED;
    free(args);
    free(timeout);
    free(command);

    if (run_status == 0) {
        httk_workflow_state_set("classification", "completed");
        httk_workflow_runlog_note("VASP completed");
        httk_workflow_advance("publish", NULL);
    } else {
        char message[96];
        snprintf(message, sizeof message, "VASP did not complete (status %d)", run_status);
        httk_workflow_fail("vasp.failed", message, NULL);
    }
    return 0;
}

static int step_publish(void) {
    int status;
    char *prefix = httk_workflow_parameter("data_prefix", "vasp", &status);
    const char *data_dir = getenv("HTTK_WORKFLOW_DATA_DIR");
    for (size_t i = 0; COLLECT[i] != NULL; i++) {
        if (!file_exists(COLLECT[i])) {
            continue;
        }
        if (data_dir != NULL && *data_dir != '\0' && prefix != NULL) {
            char *destination = join_path(prefix, COLLECT[i]);
            if (destination != NULL) {
                char *operation = httk_workflow_put(COLLECT[i], destination, &status);
                free(operation);
                free(destination);
            }
        }
    }
    httk_workflow_runlog_note(data_dir != NULL && *data_dir != '\0' ? "published to transactional data"
                                                                    : "kept the result in the workdir");
    httk_workflow_succeed();
    free(prefix);
    return 0;
}

int main(int argc, char **argv) {
    static const httk_workflow_step steps[] = {
        {"prepare", step_prepare},
        {"run", step_run},
        {"publish", step_publish},
    };
    if (httk_workflow_runner("httk.vasp.relax-c", steps, sizeof steps / sizeof steps[0]) != 0) {
        return 2;
    }
    return httk_workflow_main(argc, argv);
}

File: examples/relax_fortran/relax.f90
! One VASP relaxation, authored in modern Fortran: the same three-step shape as
! examples/relax_c/relax.c, built on the native Fortran SDK (native/fortran).
!
!   prepare  stage the payload POSCAR (and INCAR if present) into the workdir
!   run      run the configured VASP command and classify what it did
!   publish  copy the finished calculation into the job's transactional data
!
! Every httk_workflow_* call reaches the same C bridge the Python, Bash, and C
! SDKs do, so this runner is mock-vasp compatible and publishes the same bytes.
! Build it with the Makefile beside this file:
!
!     make
!
! See ../mock_vasp.py for a stand-in VASP, and README.md for the whole flow.

module relax_steps

  use, intrinsic :: iso_c_binding, only: c_int
  use, intrinsic :: iso_fortran_env, only: error_unit
  use httk_workflow

  implicit none
  private
  public :: step_prepare, step_run, step_publish

  ! The files a finished relaxation publishes, if the run produced them.
  character(len=*), parameter :: COLLECT(7) = [character(len=20) :: &
    "INCAR", "KPOINTS", "OUTCAR", "CONTCAR", "OSZICAR", "vasprun.xml", "vasp-run-report.json"]

  ! A generous fixed width for one run-argument field (a token of the command,
  ! or an option); wide enough for any realistic command path, and trailing
  ! blanks are stripped as the argument crosses the C boundary.
  integer, parameter :: ARG_WIDTH = 4096

contains

  ! One environment variable, or a default when it is unset or empty.
  function env(name, fallback) result(value)
    character(len=*), intent(in) :: name, fallback
    character(len=:), allocatable :: value
    integer :: l, s
    call get_environment_variable(name, length=l, status=s)
    if (s /= 0 .or. l == 0) then
      value = fallback
      return
    end if
    allocate (character(len=l) :: value)
    call get_environment_variable(name, value=value)
  end function

  logical function file_exists(path)
    character(len=*), intent(in) :: path
    inquire (file=path, exist=file_exists)
  end function

  ! Copy one file byte for byte; ok is .true. on success.
  subroutine copy_file(source, destination, ok)
    character(len=*), intent(in) :: source, destination
    logical, intent(out) :: ok
    character(len=:), allocatable :: buffer
    integer :: unit, ios
    integer(kind=8) :: nbytes
    ok = .false.
    inquire (file=source, size=nbytes)
    if (nbytes < 0) return
    open (newunit=unit, file=source, access="stream", form="unformatted", &
          status="old", action="read", iostat=ios)
    if (ios /= 0) return
    allocate (character(len=int(nbytes)) :: buffer)
    if (nbytes > 0) read (unit, iostat=ios) buffer
    close (unit)
    if (ios /= 0) return
    open (newunit=unit, file=destination, access="stream", form="unformatted", &
          status="replace", action="write", iostat=ios)
    if (ios /= 0) return
    if (nbytes > 0) write (unit, iostat=ios) buffer
    close (unit)
    ok = ios == 0
  end subroutine

  ! Stage a payload-relative file named by one parameter into the workdir.
  ! Returns 1 when staged, 0 when the source is absent, -1 on failure.
  function stage_input(job_dir, parameter, fallback, destination) result(result_code)
    character(len=*), intent(in) :: job_dir, parameter, fallback, destination
    integer :: result_code
    character(len=:), allocatable :: relative, source
    integer :: st
    logical :: ok
    ! The parameter has a fallback, so an OK read is always allocated; a refused
    ! bridge call (status /= OK) leaves `relative` unallocated and is a failure.
    call httk_workflow_parameter(parameter, relative, fallback, st)
    if (st /= HTTK_WORKFLOW_OK .or. .not. allocated(relative)) then
      result_code = -1
      return
    end if
    source = job_dir//"/"//relative
    if (.not. file_exists(source)) then
      result_code = 0
      return
    end if
    call copy_file(source, destination, ok)
    result_code = merge(1, -1, ok)
  end function

  function step_prepare() result(code) bind(c)
    integer(c_int) :: code
    character(len=:), allocatable :: job_dir
    integer :: staged
    job_dir = env("HTTK_WORKFLOW_JOB_DIR", ".")
    staged = stage_input(job_dir, "poscar", "files/POSCAR", "POSCAR")
    if (staged <= 0) then
      call ignore(httk_workflow_fail("vasp.input_missing", "the starting structure is not in this payload"))
      code = 0
      return
    end if
    ! An INCAR is optional; the mock VASP reads only the POSCAR.
    staged = stage_input(job_dir, "incar", "files/INCAR", "INCAR")
    call ignore(httk_workflow_runlog_note("prepared a relaxation"))
    call ignore(httk_workflow_advance("run"))
    code = 0
  end function

  function step_run() result(code) bind(c)
    integer(c_int) :: code
    character(len=:), allocatable :: from_parameter, command, timeout
    character(len=ARG_WIDTH), allocatable :: args(:)
    character(len=64) :: message
    integer :: run_status
    call httk_workflow_parameter("vasp_command", from_parameter, "")
    if (.not. allocated(from_parameter)) from_parameter = ""
    call httk_workflow_setting("vasp.command", command, from_parameter)
    if (.not. allocated(command)) command = ""
    if (len_trim(command) == 0) then
      call ignore(httk_workflow_fail("vasp.command_missing", &
        "no VASP command is configured: set it with "// &
        "httk workflow workspace settings set vasp.command '...', or set "// &
        "HTTK_VASP_COMMAND, or give the job a vasp_command parameter"))
      code = 0
      return
    end if
    call httk_workflow_parameter("timeout", timeout, "86400")
    if (.not. allocated(timeout)) timeout = "86400"
    call build_run_args(timeout, command, args)
    run_status = httk_workflow_run(args)
    if (run_status == 0) then
      call ignore(httk_workflow_state_set("classification", "completed"))
      call ignore(httk_workflow_runlog_note("VASP completed"))
      call ignore(httk_workflow_advance("publish"))
    else
      write (message, '(A,I0,A)') "VASP did not complete (status ", run_status, ")"
      call ignore(httk_workflow_fail("vasp.failed", trim(message)))
    end if
    code = 0
  end function

  ! Build the tail httk_workflow_run forwards: the fixed options, then "--", then
  ! the configured command word-split on whitespace, as the Bash runner leaves it
  ! unquoted for the shell.
  subroutine build_run_args(timeout, command, args)
    character(len=*), intent(in) :: timeout, command
    character(len=ARG_WIDTH), allocatable, intent(out) :: args(:)
    integer, parameter :: FIXED = 5
    integer :: i, n, start, ntok
    integer, allocatable :: tstart(:), tend(:)
    logical :: in_token
    ! One pass to find the token boundaries of the command.
    n = len_trim(command)
    allocate (tstart(n), tend(n))
    ntok = 0
    in_token = .false.
    start = 0
    do i = 1, n
      if (is_space(command(i:i))) then
        if (in_token) then
          ntok = ntok + 1
          tstart(ntok) = start
          tend(ntok) = i - 1
          in_token = .false.
        end if
      else
        if (.not. in_token) then
          start = i
          in_token = .true.
        end if
      end if
    end do
    if (in_token) then
      ntok = ntok + 1
      tstart(ntok) = start
      tend(ntok) = n
    end if
    allocate (args(FIXED + ntok))
    args(1) = "--timeout"
    args(2) = timeout
    args(3) = "--report"
    args(4) = "vasp-run-report.json"
    args(5) = "--"
    do i = 1, ntok
      ! Loud failure rather than silent truncation: a token wider than the fixed
      ! ARG_WIDTH field would be clipped, running the wrong command.
      if (tend(i) - tstart(i) + 1 > ARG_WIDTH) then
        write (error_unit, '(A,I0,A)') "relax: a command token exceeds ARG_WIDTH (", ARG_WIDTH, ")"
        call httk_workflow_exit(2)
      end if
      args(FIXED + i) = command(tstart(i):tend(i))
    end do
  end subroutine

  logical function is_space(c)
    character, intent(in) :: c
    is_space = c == " " .or. iachar(c) == 9
  end function

  function step_publish() result(code) bind(c)
    integer(c_int) :: code
    character(len=:), allocatable :: prefix, data_dir, operation
    integer :: i
    logical :: to_data
    call httk_workflow_parameter("data_prefix", prefix, "vasp")
    if (.not. allocated(prefix)) prefix = "vasp"
    data_dir = env("HTTK_WORKFLOW_DATA_DIR", "")
    to_data = len_trim(data_dir) > 0
    do i = 1, size(COLLECT)
      if (.not. file_exists(trim(COLLECT(i)))) cycle
      if (to_data) then
        call httk_workflow_put(trim(COLLECT(i)), trim(prefix)//"/"//trim(COLLECT(i)), operation)
      end if
    end do
    if (to_data) then
      call ignore(httk_workflow_runlog_note("published to transactional data"))
    else
      call ignore(httk_workflow_runlog_note("kept the result in the workdir"))
    end if
    call ignore(httk_workflow_succeed())
    code = 0
  end function

end module relax_steps

program relax

  use, intrinsic :: iso_c_binding, only: c_funloc
  use httk_workflow
  use relax_steps

  implicit none

  if (httk_workflow_runner("httk.vasp.relax-fortran", &
        [character(len=8) :: "prepare", "run", "publish"], &
        [c_funloc(step_prepare), c_funloc(step_run), c_funloc(step_publish)]) /= HTTK_WORKFLOW_OK) then
    call httk_workflow_exit(2)
  end if
  call httk_workflow_exit(httk_workflow_main())

end program relax
